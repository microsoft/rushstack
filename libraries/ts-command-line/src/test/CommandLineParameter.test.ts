// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as argparse from 'argparse';
import { AnsiEscape } from '@rushstack/terminal';

import { DynamicCommandLineParser } from '../providers/DynamicCommandLineParser';
import { DynamicCommandLineAction } from '../providers/DynamicCommandLineAction';
import { CommandLineParameterBase } from '../parameters/BaseClasses';
import type { CommandLineParser } from '../providers/CommandLineParser';
import type { CommandLineAction } from '../providers/CommandLineAction';

interface IExtendedArgumentParser extends argparse.ArgumentParser {
  _printMessage: (message: string) => void;
}

function createParser(): DynamicCommandLineParser {
  const commandLineParser: DynamicCommandLineParser = new DynamicCommandLineParser({
    toolFilename: 'example',
    toolDescription: 'An example project'
  });
  commandLineParser.defineFlagParameter({
    parameterLongName: '--global-flag',
    parameterShortName: '-g',
    description: 'A flag that affects all actions'
  });

  const action: DynamicCommandLineAction = new DynamicCommandLineAction({
    actionName: 'do:the-job',
    summary: 'does the job',
    documentation: 'a longer description'
  });
  commandLineParser.addAction(action);

  // Choice
  action.defineChoiceParameter({
    parameterLongName: '--choice',
    parameterShortName: '-c',
    description: 'A choice',
    alternatives: ['one', 'two', 'three', 'default'],
    environmentVariable: 'ENV_CHOICE'
  });
  action.defineChoiceParameter({
    parameterLongName: '--choice-with-default',
    description: 'A choice with a default. This description ends with a "quoted word"',
    alternatives: ['one', 'two', 'three', 'default'],
    environmentVariable: 'ENV_CHOICE2',
    defaultValue: 'default'
  });

  // Choice List
  action.defineChoiceListParameter({
    parameterLongName: '--choice-list',
    parameterShortName: '-C',
    description: 'This parameter may be specified multiple times to make a list of choices',
    alternatives: ['red', 'green', 'blue'],
    environmentVariable: 'ENV_CHOICE_LIST'
  });

  // Flag
  action.defineFlagParameter({
    parameterLongName: '--flag',
    parameterShortName: '-f',
    description: 'A flag',
    environmentVariable: 'ENV_FLAG'
  });

  // Integer
  action.defineIntegerParameter({
    parameterLongName: '--integer',
    parameterShortName: '-i',
    description: 'An integer',
    argumentName: 'NUMBER',
    environmentVariable: 'ENV_INTEGER'
  });
  action.defineIntegerParameter({
    parameterLongName: '--integer-with-default',
    description: 'An integer with a default',
    argumentName: 'NUMBER',
    environmentVariable: 'ENV_INTEGER2',
    defaultValue: 123
  });
  action.defineIntegerParameter({
    parameterLongName: '--integer-required',
    description: 'An integer',
    argumentName: 'NUMBER',
    required: true
  });
  action.defineIntegerParameter({
    parameterLongName: '--env-integer-required',
    description: 'An integer',
    argumentName: 'NUMBER',
    environmentVariable: 'ENV_INTEGER_REQUIRED',
    required: true
  });

  // Integer List
  action.defineIntegerListParameter({
    parameterLongName: '--integer-list',
    parameterShortName: '-I',
    description: 'This parameter may be specified multiple times to make a list of integers',
    argumentName: 'LIST_ITEM',
    environmentVariable: 'ENV_INTEGER_LIST'
  });

  // String
  action.defineStringParameter({
    parameterLongName: '--string',
    parameterShortName: '-s',
    description: 'A string',
    argumentName: 'TEXT',
    environmentVariable: 'ENV_STRING'
  });
  action.defineStringParameter({
    parameterLongName: '--string-with-default',
    description: 'A string with a default',
    argumentName: 'TEXT',
    environmentVariable: 'ENV_STRING2',
    defaultValue: '123'
  });
  action.defineStringParameter({
    parameterLongName: '--string-with-undocumented-synonym',
    description: 'A string with an undocumented synonym',
    argumentName: 'TEXT',
    undocumentedSynonyms: ['--undocumented-synonym']
  });

  // String List
  action.defineStringListParameter({
    parameterLongName: '--string-list',
    parameterShortName: '-l',
    description: 'This parameter may be specified multiple times to make a list of strings',
    argumentName: 'LIST_ITEM',
    environmentVariable: 'ENV_STRING_LIST'
  });

  return commandLineParser;
}

function expectPropertiesToMatchSnapshot(object: {}, propertyNames: string[]): void {
  const snapshotObject: {} = {};

  for (const propertyName of propertyNames) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (snapshotObject as any)[propertyName] = (object as any)[propertyName];
  }
  expect(snapshotObject).toMatchSnapshot();
}

const snapshotPropertyNames: string[] = [
  'description',
  'kind',
  'longName',
  'shortName',
  'value',
  'kind',
  'argumentName',
  'environmentVariable',
  'required',
  'defaultValue',
  'values'
];

describe(CommandLineParameterBase.name, () => {
  let existingEnv: NodeJS.ProcessEnv;

  beforeEach(() => {
    existingEnv = {
      ...process.env
    };
  });

  afterEach(() => {
    process.env = existingEnv;
  });

  it('prints the global help', () => {
    const commandLineParser: CommandLineParser = createParser();
    const helpText: string = AnsiEscape.removeCodes(commandLineParser.renderHelpText());
    expect(helpText).toMatchSnapshot();
  });

  it('prints the action help', () => {
    const commandLineParser: CommandLineParser = createParser();
    const helpText: string = AnsiEscape.removeCodes(
      commandLineParser.getAction('do:the-job').renderHelpText()
    );
    expect(helpText).toMatchSnapshot();
  });

  it('parses an input with ALL parameters', async () => {
    const commandLineParser: CommandLineParser = createParser();
    const action: CommandLineAction = commandLineParser.getAction('do:the-job');

    const args: string[] = [
      '--global-flag',
      'do:the-job',
      '--choice',
      'two',
      '--choice-list',
      'red',
      '--choice-list',
      'blue',
      '--flag',
      '--integer',
      '123',
      '--integer-required',
      '321',
      '--env-integer-required',
      '123',
      '--integer-list',
      '37',
      '--integer-list',
      '-404',
      '--string',
      'hello',
      '--string-list',
      'first',
      '--string-list',
      'second'
    ];

    await expect(commandLineParser.executeAsync(args)).resolves.toBe(true);

    expect(commandLineParser.selectedAction).toBe(action);

    expectPropertiesToMatchSnapshot(
      commandLineParser.getFlagParameter('--global-flag'),
      snapshotPropertyNames
    );

    expectPropertiesToMatchSnapshot(action.getChoiceParameter('--choice'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(
      action.getChoiceParameter('--choice-with-default'),
      snapshotPropertyNames
    );
    expectPropertiesToMatchSnapshot(action.getChoiceListParameter('--choice-list'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(action.getFlagParameter('--flag'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(action.getIntegerParameter('--integer'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(
      action.getIntegerParameter('--integer-with-default'),
      snapshotPropertyNames
    );
    expectPropertiesToMatchSnapshot(action.getIntegerParameter('--integer-required'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(action.getIntegerListParameter('--integer-list'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(action.getStringParameter('--string'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(
      action.getStringParameter('--string-with-default'),
      snapshotPropertyNames
    );
    expectPropertiesToMatchSnapshot(action.getStringListParameter('--string-list'), snapshotPropertyNames);

    const copiedArgs: string[] = [];
    for (const parameter of action.parameters) {
      copiedArgs.push(`### ${parameter.longName} output: ###`);
      parameter.appendToArgList(copiedArgs);
    }
    expect(copiedArgs).toMatchSnapshot();
  });

  it('parses an input with NO parameters', async () => {
    const commandLineParser: CommandLineParser = createParser();
    const action: CommandLineAction = commandLineParser.getAction('do:the-job');
    const args: string[] = ['do:the-job', '--integer-required', '123', '--env-integer-required', '321'];

    await expect(commandLineParser.executeAsync(args)).resolves.toBe(true);

    expect(commandLineParser.selectedAction).toBe(action);

    expectPropertiesToMatchSnapshot(
      commandLineParser.getFlagParameter('--global-flag'),
      snapshotPropertyNames
    );

    expectPropertiesToMatchSnapshot(action.getChoiceParameter('--choice'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(
      action.getChoiceParameter('--choice-with-default'),
      snapshotPropertyNames
    );
    expectPropertiesToMatchSnapshot(action.getChoiceListParameter('--choice-list'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(action.getFlagParameter('--flag'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(action.getIntegerParameter('--integer'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(
      action.getIntegerParameter('--integer-with-default'),
      snapshotPropertyNames
    );
    expectPropertiesToMatchSnapshot(action.getIntegerParameter('--integer-required'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(action.getIntegerListParameter('--integer-list'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(action.getStringParameter('--string'), snapshotPropertyNames);
    expectPropertiesToMatchSnapshot(
      action.getStringParameter('--string-with-default'),
      snapshotPropertyNames
    );
    expectPropertiesToMatchSnapshot(action.getStringListParameter('--string-list'), snapshotPropertyNames);

    const copiedArgs: string[] = [];
    for (const parameter of action.parameters) {
      copiedArgs.push(`### ${parameter.longName} output: ###`);
      parameter.appendToArgList(copiedArgs);
    }
    expect(copiedArgs).toMatchSnapshot();
  });

  it('parses each parameter from an environment variable', async () => {
    const commandLineParser: CommandLineParser = createParser();
    const action: CommandLineAction = commandLineParser.getAction('do:the-job');

    action.defineStringListParameter({
      parameterLongName: '--json-string-list',
      description: 'Test JSON parsing',
      argumentName: 'LIST_ITEM',
      environmentVariable: 'ENV_JSON_STRING_LIST'
    });

    const args: string[] = ['do:the-job', '--integer-required', '1'];

    process.env.ENV_CHOICE = 'one';
    process.env.ENV_CHOICE2 = 'two';
    process.env.ENV_CHOICE_LIST = ' [ "red", "green" ] ';
    process.env.ENV_FLAG = '1';
    process.env.ENV_INTEGER = '111';
    process.env.ENV_INTEGER2 = '222';
    process.env.ENV_INTEGER_REQUIRED = '333';
    process.env.ENV_INTEGER_LIST = ' [ 1 , 2 , 3 ] ';
    process.env.ENV_STRING = 'Hello, world!';
    process.env.ENV_STRING2 = 'Hello, world!';
    process.env.ENV_STRING_LIST = 'simple text';
    process.env.ENV_JSON_STRING_LIST = ' [ 1, true, "Hello, world!" ] ';

    await expect(commandLineParser.executeAsync(args)).resolves.toBe(true);

    expect(commandLineParser.selectedAction).toBe(action);

    const copiedArgs: string[] = [];
    for (const parameter of action.parameters) {
      copiedArgs.push(`### ${parameter.longName} output: ###`);
      parameter.appendToArgList(copiedArgs);
    }
    expect(copiedArgs).toMatchSnapshot();
  });

  it('allows an undocumented synonym', async () => {
    const commandLineParser: CommandLineParser = createParser();
    const action: CommandLineAction = commandLineParser.getAction('do:the-job');

    const args: string[] = [
      'do:the-job',
      '--undocumented-synonym',
      'undocumented-value',
      '--integer-required',
      '6',
      '--env-integer-required',
      '123'
    ];

    await expect(commandLineParser.executeAsync(args)).resolves.toBe(true);

    expect(commandLineParser.selectedAction).toBe(action);

    const copiedArgs: string[] = [];
    for (const parameter of action.parameters) {
      copiedArgs.push(`### ${parameter.longName} output: ###`);
      parameter.appendToArgList(copiedArgs);
    }
    expect(copiedArgs).toMatchSnapshot();
  });

  it('raises an error if a required parameter backed by an env variable is not provided', async () => {
    const commandLineParser: CommandLineParser = createParser();

    const printMessageSpy: jest.SpyInstance = jest
      .spyOn(argparse.ArgumentParser.prototype as IExtendedArgumentParser, '_printMessage')
      .mockImplementation(() => {
        /* don't print */
      });

    const args: string[] = ['do:the-job', '--integer-required', '1'];
    await expect(commandLineParser.executeWithoutErrorHandlingAsync(args)).rejects.toMatchSnapshot('Error');
    expect(printMessageSpy).toHaveBeenCalled();
    expect(printMessageSpy.mock.calls[0][0]).toMatchSnapshot('Usage');
  });

  it(
    'prints the same usage if a required parameter backed by an env variable is not provided as when ' +
      'a different required parameter is missing',
    async () => {
      const printMessageSpy: jest.SpyInstance = jest
        .spyOn(argparse.ArgumentParser.prototype as IExtendedArgumentParser, '_printMessage')
        .mockImplementation(() => {
          /* don't print */
        });

      async function runWithArgsAsync(args: string[]): Promise<void> {
        const commandLineParser: CommandLineParser = createParser();
        await expect(commandLineParser.executeAsync(args)).resolves.toBe(false);
      }

      await runWithArgsAsync(['do:the-job', '--integer-required', '1']);
      await runWithArgsAsync(['do:the-job', '--env-integer-required', '1']);

      expect(printMessageSpy).toHaveBeenCalledTimes(2);
      expect(printMessageSpy.mock.calls[0][0]).toMatchSnapshot('Usage');
      expect(printMessageSpy.mock.calls[0][0]).toEqual(printMessageSpy.mock.calls[1][0]);
    }
  );

  describe('choice list', () => {
    function createHelloWorldParser(): CommandLineParser {
      const commandLineParser: CommandLineParser = new DynamicCommandLineParser({
        toolFilename: 'example',
        toolDescription: 'An example project'
      });
      const action: DynamicCommandLineAction = new DynamicCommandLineAction({
        actionName: 'hello-world',
        summary: 'Hello World',
        documentation: 'best program'
      });
      commandLineParser.addAction(action);

      action.defineChoiceListParameter({
        parameterLongName: '--color',
        parameterShortName: '-c',
        description: 'Your favorite colors',
        alternatives: ['purple', 'yellow', 'pizza'],
        environmentVariable: 'ENV_COLOR'
      });

      return commandLineParser;
    }

    it('raises an error if env var value is not valid json', async () => {
      const commandLineParser: CommandLineParser = createHelloWorldParser();
      const args: string[] = ['hello-world'];
      process.env.ENV_COLOR = '[u';

      // TODO: When Node 18 support is removed, switch this to use
      // ```
      // await expect(
      //   commandLineParser.executeWithoutErrorHandling(args)
      // ).rejects.toThrowErrorMatchingSnapshot();
      // ```

      let error: string | undefined;
      try {
        await commandLineParser.executeWithoutErrorHandlingAsync(args);
      } catch (e) {
        error = e.message;
      }

      expect(error).toMatch(
        /^The \[u environment variable value looks like a JSON array but failed to parse: Unexpected token /
      );
    });

    it('raises an error if env var value is json containing non-scalars', async () => {
      const commandLineParser: CommandLineParser = createHelloWorldParser();
      const args: string[] = ['hello-world'];
      process.env.ENV_COLOR = '[{}]';

      await expect(
        commandLineParser.executeWithoutErrorHandlingAsync(args)
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('raises an error if env var value is not a valid choice', async () => {
      const commandLineParser: CommandLineParser = createHelloWorldParser();
      const args: string[] = ['hello-world'];
      process.env.ENV_COLOR = 'oblong';

      await expect(
        commandLineParser.executeWithoutErrorHandlingAsync(args)
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
