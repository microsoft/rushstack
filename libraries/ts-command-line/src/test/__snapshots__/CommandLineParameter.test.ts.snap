// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandLineParameterBase allows an undocumented synonym 1`] = `
Array [
  "### --choice output: ###",
  "### --choice-with-default output: ###",
  "--choice-with-default",
  "default",
  "### --choice-list output: ###",
  "### --flag output: ###",
  "### --integer output: ###",
  "### --integer-with-default output: ###",
  "--integer-with-default",
  "123",
  "### --integer-required output: ###",
  "--integer-required",
  "6",
  "### --env-integer-required output: ###",
  "--env-integer-required",
  "123",
  "### --integer-list output: ###",
  "### --string output: ###",
  "### --string-with-default output: ###",
  "--string-with-default",
  "123",
  "### --string-with-undocumented-synonym output: ###",
  "--string-with-undocumented-synonym",
  "undocumented-value",
  "### --string-list output: ###",
]
`;

exports[`CommandLineParameterBase choice list raises an error if env var value is json containing non-scalars 1`] = `"The [{}] environment variable value looks like a JSON array but failed to parse: The [{}] environment variable value must be a JSON  array containing only strings, numbers, and booleans."`;

exports[`CommandLineParameterBase choice list raises an error if env var value is not a valid choice 1`] = `"Invalid value \\"oblong\\" for the environment variable ENV_COLOR.  Valid choices are: \\"purple\\", \\"yellow\\", \\"pizza\\""`;

exports[`CommandLineParameterBase parses an input with ALL parameters 1`] = `
Object {
  "argumentName": undefined,
  "defaultValue": undefined,
  "description": "A flag that affects all actions",
  "environmentVariable": undefined,
  "kind": 1,
  "longName": "--global-flag",
  "required": false,
  "shortName": "-g",
  "value": true,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 2`] = `
Object {
  "argumentName": undefined,
  "defaultValue": undefined,
  "description": "A choice",
  "environmentVariable": "ENV_CHOICE",
  "kind": -0,
  "longName": "--choice",
  "required": false,
  "shortName": "-c",
  "value": "two",
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 3`] = `
Object {
  "argumentName": undefined,
  "defaultValue": "default",
  "description": "A choice with a default. This description ends with a \\"quoted word\\"",
  "environmentVariable": "ENV_CHOICE2",
  "kind": -0,
  "longName": "--choice-with-default",
  "required": false,
  "shortName": undefined,
  "value": "default",
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 4`] = `
Object {
  "argumentName": undefined,
  "defaultValue": undefined,
  "description": "This parameter may be specified multiple times to make a list of choices",
  "environmentVariable": "ENV_CHOICE_LIST",
  "kind": 5,
  "longName": "--choice-list",
  "required": false,
  "shortName": "-C",
  "value": undefined,
  "values": Array [
    "red",
    "blue",
  ],
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 5`] = `
Object {
  "argumentName": undefined,
  "defaultValue": undefined,
  "description": "A flag",
  "environmentVariable": "ENV_FLAG",
  "kind": 1,
  "longName": "--flag",
  "required": false,
  "shortName": "-f",
  "value": true,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 6`] = `
Object {
  "argumentName": "NUMBER",
  "defaultValue": undefined,
  "description": "An integer",
  "environmentVariable": "ENV_INTEGER",
  "kind": 2,
  "longName": "--integer",
  "required": false,
  "shortName": "-i",
  "value": 123,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 7`] = `
Object {
  "argumentName": "NUMBER",
  "defaultValue": 123,
  "description": "An integer with a default",
  "environmentVariable": "ENV_INTEGER2",
  "kind": 2,
  "longName": "--integer-with-default",
  "required": false,
  "shortName": undefined,
  "value": 123,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 8`] = `
Object {
  "argumentName": "NUMBER",
  "defaultValue": undefined,
  "description": "An integer",
  "environmentVariable": undefined,
  "kind": 2,
  "longName": "--integer-required",
  "required": true,
  "shortName": undefined,
  "value": 321,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 9`] = `
Object {
  "argumentName": "LIST_ITEM",
  "defaultValue": undefined,
  "description": "This parameter may be specified multiple times to make a list of integers",
  "environmentVariable": "ENV_INTEGER_LIST",
  "kind": 6,
  "longName": "--integer-list",
  "required": false,
  "shortName": "-I",
  "value": undefined,
  "values": Array [
    37,
    -404,
  ],
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 10`] = `
Object {
  "argumentName": "TEXT",
  "defaultValue": undefined,
  "description": "A string",
  "environmentVariable": "ENV_STRING",
  "kind": 3,
  "longName": "--string",
  "required": false,
  "shortName": "-s",
  "value": "hello",
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 11`] = `
Object {
  "argumentName": "TEXT",
  "defaultValue": "123",
  "description": "A string with a default",
  "environmentVariable": "ENV_STRING2",
  "kind": 3,
  "longName": "--string-with-default",
  "required": false,
  "shortName": undefined,
  "value": "123",
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 12`] = `
Object {
  "argumentName": "LIST_ITEM",
  "defaultValue": undefined,
  "description": "This parameter may be specified multiple times to make a list of strings",
  "environmentVariable": "ENV_STRING_LIST",
  "kind": 4,
  "longName": "--string-list",
  "required": false,
  "shortName": "-l",
  "value": undefined,
  "values": Array [
    "first",
    "second",
  ],
}
`;

exports[`CommandLineParameterBase parses an input with ALL parameters 13`] = `
Array [
  "### --choice output: ###",
  "--choice",
  "two",
  "### --choice-with-default output: ###",
  "--choice-with-default",
  "default",
  "### --choice-list output: ###",
  "--choice-list",
  "red",
  "--choice-list",
  "blue",
  "### --flag output: ###",
  "--flag",
  "### --integer output: ###",
  "--integer",
  "123",
  "### --integer-with-default output: ###",
  "--integer-with-default",
  "123",
  "### --integer-required output: ###",
  "--integer-required",
  "321",
  "### --env-integer-required output: ###",
  "--env-integer-required",
  "123",
  "### --integer-list output: ###",
  "--integer-list",
  "37",
  "--integer-list",
  "-404",
  "### --string output: ###",
  "--string",
  "hello",
  "### --string-with-default output: ###",
  "--string-with-default",
  "123",
  "### --string-with-undocumented-synonym output: ###",
  "### --string-list output: ###",
  "--string-list",
  "first",
  "--string-list",
  "second",
]
`;

exports[`CommandLineParameterBase parses an input with NO parameters 1`] = `
Object {
  "argumentName": undefined,
  "defaultValue": undefined,
  "description": "A flag that affects all actions",
  "environmentVariable": undefined,
  "kind": 1,
  "longName": "--global-flag",
  "required": false,
  "shortName": "-g",
  "value": false,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 2`] = `
Object {
  "argumentName": undefined,
  "defaultValue": undefined,
  "description": "A choice",
  "environmentVariable": "ENV_CHOICE",
  "kind": -0,
  "longName": "--choice",
  "required": false,
  "shortName": "-c",
  "value": undefined,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 3`] = `
Object {
  "argumentName": undefined,
  "defaultValue": "default",
  "description": "A choice with a default. This description ends with a \\"quoted word\\"",
  "environmentVariable": "ENV_CHOICE2",
  "kind": -0,
  "longName": "--choice-with-default",
  "required": false,
  "shortName": undefined,
  "value": "default",
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 4`] = `
Object {
  "argumentName": undefined,
  "defaultValue": undefined,
  "description": "This parameter may be specified multiple times to make a list of choices",
  "environmentVariable": "ENV_CHOICE_LIST",
  "kind": 5,
  "longName": "--choice-list",
  "required": false,
  "shortName": "-C",
  "value": undefined,
  "values": Array [],
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 5`] = `
Object {
  "argumentName": undefined,
  "defaultValue": undefined,
  "description": "A flag",
  "environmentVariable": "ENV_FLAG",
  "kind": 1,
  "longName": "--flag",
  "required": false,
  "shortName": "-f",
  "value": false,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 6`] = `
Object {
  "argumentName": "NUMBER",
  "defaultValue": undefined,
  "description": "An integer",
  "environmentVariable": "ENV_INTEGER",
  "kind": 2,
  "longName": "--integer",
  "required": false,
  "shortName": "-i",
  "value": undefined,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 7`] = `
Object {
  "argumentName": "NUMBER",
  "defaultValue": 123,
  "description": "An integer with a default",
  "environmentVariable": "ENV_INTEGER2",
  "kind": 2,
  "longName": "--integer-with-default",
  "required": false,
  "shortName": undefined,
  "value": 123,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 8`] = `
Object {
  "argumentName": "NUMBER",
  "defaultValue": undefined,
  "description": "An integer",
  "environmentVariable": undefined,
  "kind": 2,
  "longName": "--integer-required",
  "required": true,
  "shortName": undefined,
  "value": 123,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 9`] = `
Object {
  "argumentName": "LIST_ITEM",
  "defaultValue": undefined,
  "description": "This parameter may be specified multiple times to make a list of integers",
  "environmentVariable": "ENV_INTEGER_LIST",
  "kind": 6,
  "longName": "--integer-list",
  "required": false,
  "shortName": "-I",
  "value": undefined,
  "values": Array [],
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 10`] = `
Object {
  "argumentName": "TEXT",
  "defaultValue": undefined,
  "description": "A string",
  "environmentVariable": "ENV_STRING",
  "kind": 3,
  "longName": "--string",
  "required": false,
  "shortName": "-s",
  "value": undefined,
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 11`] = `
Object {
  "argumentName": "TEXT",
  "defaultValue": "123",
  "description": "A string with a default",
  "environmentVariable": "ENV_STRING2",
  "kind": 3,
  "longName": "--string-with-default",
  "required": false,
  "shortName": undefined,
  "value": "123",
  "values": undefined,
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 12`] = `
Object {
  "argumentName": "LIST_ITEM",
  "defaultValue": undefined,
  "description": "This parameter may be specified multiple times to make a list of strings",
  "environmentVariable": "ENV_STRING_LIST",
  "kind": 4,
  "longName": "--string-list",
  "required": false,
  "shortName": "-l",
  "value": undefined,
  "values": Array [],
}
`;

exports[`CommandLineParameterBase parses an input with NO parameters 13`] = `
Array [
  "### --choice output: ###",
  "### --choice-with-default output: ###",
  "--choice-with-default",
  "default",
  "### --choice-list output: ###",
  "### --flag output: ###",
  "### --integer output: ###",
  "### --integer-with-default output: ###",
  "--integer-with-default",
  "123",
  "### --integer-required output: ###",
  "--integer-required",
  "123",
  "### --env-integer-required output: ###",
  "--env-integer-required",
  "321",
  "### --integer-list output: ###",
  "### --string output: ###",
  "### --string-with-default output: ###",
  "--string-with-default",
  "123",
  "### --string-with-undocumented-synonym output: ###",
  "### --string-list output: ###",
]
`;

exports[`CommandLineParameterBase parses each parameter from an environment variable 1`] = `
Array [
  "### --choice output: ###",
  "--choice",
  "one",
  "### --choice-with-default output: ###",
  "--choice-with-default",
  "two",
  "### --choice-list output: ###",
  "--choice-list",
  "red",
  "--choice-list",
  "green",
  "### --flag output: ###",
  "--flag",
  "### --integer output: ###",
  "--integer",
  "111",
  "### --integer-with-default output: ###",
  "--integer-with-default",
  "222",
  "### --integer-required output: ###",
  "--integer-required",
  "1",
  "### --env-integer-required output: ###",
  "--env-integer-required",
  "333",
  "### --integer-list output: ###",
  "--integer-list",
  "1",
  "--integer-list",
  "2",
  "--integer-list",
  "3",
  "### --string output: ###",
  "--string",
  "Hello, world!",
  "### --string-with-default output: ###",
  "--string-with-default",
  "Hello, world!",
  "### --string-with-undocumented-synonym output: ###",
  "### --string-list output: ###",
  "--string-list",
  "simple text",
  "### --json-string-list output: ###",
  "--json-string-list",
  "1",
  "--json-string-list",
  "true",
  "--json-string-list",
  "Hello, world!",
]
`;

exports[`CommandLineParameterBase prints the action help 1`] = `
"usage: example do:the-job [-h] [-c {one,two,three,default}]
                          [--choice-with-default {one,two,three,default}]
                          [-C {red,green,blue}] [-f] [-i NUMBER]
                          [--integer-with-default NUMBER] --integer-required
                          NUMBER --env-integer-required NUMBER [-I LIST_ITEM]
                          [-s TEXT] [--string-with-default TEXT]
                          [--string-with-undocumented-synonym TEXT]
                          [-l LIST_ITEM]
                          

a longer description

Optional arguments:
  -h, --help            Show this help message and exit.
  -c {one,two,three,default}, --choice {one,two,three,default}
                        A choice. This parameter may alternatively be 
                        specified via the ENV_CHOICE environment variable.
  --choice-with-default {one,two,three,default}
                        A choice with a default. This description ends with a 
                        \\"quoted word\\". This parameter may alternatively be 
                        specified via the ENV_CHOICE2 environment variable. 
                        The default value is \\"default\\".
  -C {red,green,blue}, --choice-list {red,green,blue}
                        This parameter may be specified multiple times to 
                        make a list of choices. This parameter may 
                        alternatively be specified via the ENV_CHOICE_LIST 
                        environment variable.
  -f, --flag            A flag. This parameter may alternatively be specified 
                        via the ENV_FLAG environment variable.
  -i NUMBER, --integer NUMBER
                        An integer. This parameter may alternatively be 
                        specified via the ENV_INTEGER environment variable.
  --integer-with-default NUMBER
                        An integer with a default. This parameter may 
                        alternatively be specified via the ENV_INTEGER2 
                        environment variable. The default value is 123.
  --integer-required NUMBER
                        An integer
  --env-integer-required NUMBER
                        An integer. This parameter may alternatively be 
                        specified via the ENV_INTEGER_REQUIRED environment 
                        variable.
  -I LIST_ITEM, --integer-list LIST_ITEM
                        This parameter may be specified multiple times to 
                        make a list of integers. This parameter may 
                        alternatively be specified via the ENV_INTEGER_LIST 
                        environment variable.
  -s TEXT, --string TEXT
                        A string. This parameter may alternatively be 
                        specified via the ENV_STRING environment variable.
  --string-with-default TEXT
                        A string with a default. This parameter may 
                        alternatively be specified via the ENV_STRING2 
                        environment variable. The default value is \\"123\\".
  --string-with-undocumented-synonym TEXT
                        A string with an undocumented synonym
  -l LIST_ITEM, --string-list LIST_ITEM
                        This parameter may be specified multiple times to 
                        make a list of strings. This parameter may 
                        alternatively be specified via the ENV_STRING_LIST 
                        environment variable.
"
`;

exports[`CommandLineParameterBase prints the global help 1`] = `
"usage: example [-h] [-g] <command> ...

An example project

Positional arguments:
  <command>
    do:the-job       does the job

Optional arguments:
  -h, --help         Show this help message and exit.
  -g, --global-flag  A flag that affects all actions

For detailed help about a specific command, use: example <command> -h
"
`;

exports[`CommandLineParameterBase prints the same usage if a required parameter backed by an env variable is not provided as when a different required parameter is missing: Usage 1`] = `
"usage: example do:the-job [-h] [-c {one,two,three,default}]
                          [--choice-with-default {one,two,three,default}]
                          [-C {red,green,blue}] [-f] [-i NUMBER]
                          [--integer-with-default NUMBER] --integer-required
                          NUMBER --env-integer-required NUMBER [-I LIST_ITEM]
                          [-s TEXT] [--string-with-default TEXT]
                          [--string-with-undocumented-synonym TEXT]
                          [-l LIST_ITEM]
                          
"
`;

exports[`CommandLineParameterBase raises an error if a required parameter backed by an env variable is not provided: Error 1`] = `
[Error: example do:the-job: error: Argument "--env-integer-required" is required
]
`;

exports[`CommandLineParameterBase raises an error if a required parameter backed by an env variable is not provided: Usage 1`] = `
"usage: example do:the-job [-h] [-c {one,two,three,default}]
                          [--choice-with-default {one,two,three,default}]
                          [-C {red,green,blue}] [-f] [-i NUMBER]
                          [--integer-with-default NUMBER] --integer-required
                          NUMBER --env-integer-required NUMBER [-I LIST_ITEM]
                          [-s TEXT] [--string-with-default TEXT]
                          [--string-with-undocumented-synonym TEXT]
                          [-l LIST_ITEM]
                          
"
`;
